dali/kernels/dynamic_scratchpad.h

```diff
@@ -0,0 +1,170 @@
+  template <typename Kind>
+  void set_upstream_resrouce(mm::memory_resource<Kind> *rsrc) {
+    resource<Kind>() = mm::monotonic_memory_resource<Kind>(rsrc, initial_size<Kind>());
+  }
+
+  template <typename Kind>
+  void set_upstream_resrouce(mm::async_memory_resource<Kind> *rsrc,
+                             AccessOrder alloc_order,
+                             AccessOrder dealloc_order = {}) {
+    static_assert(!std::is_same<Kind, mm::memory_kind::host>::value,
+      "Cannot use a stream-ordered resource for plain host memory");
+    adapter<Kind>() = { rsrc, alloc_order, dealloc_order };
+    set_upstream_resrouce<Kind>(&adapter<Kind>());
+  }
+
+  template <typename Kind>
+  size_t &initial_size() {
+    return initial_sizes_[index_in_pack<Kind, Kinds...>::value];
+  }
+
+  template <typename Kind>
+  size_t initial_size() const {
+    return initial_sizes_[index_in_pack<Kind, Kinds...>::value];
+  }
+
+  template <typename Kind>
+  mm::memory_resource<Kind> *get_upstream() const {
+    std::get<mm::monotonic_memory_resource<Kind>>(resources_)->get_upstream();
+  }
+
+  template <typename Kind>
+  auto &adapter() {
+    return std::get<mm::fixed_order_resource<Kind>>(adapters_);
+  }
+
+  template <typename Kind>
+  auto &adapter() const {
+    return std::get<mm::fixed_order_resource<Kind>>(adapters_);
+  }
+
+  template <typename Kind>
+  auto &resource() {
+    return std::get<mm::monotonic_memory_resource<Kind>>(resources_);
+  }
+
+  template <typename Kind>
+  auto &resource() const {
+    return std::get<mm::monotonic_memory_resource<Kind>>(resources_);
+  }
+
+  std::tuple<mm::fixed_order_resource<Kinds>...>      adapters_;
+  std::tuple<mm::monotonic_memory_resource<Kinds>...> resources_;
+  std::array<size_t, sizeof...(Kinds)>                initial_sizes_ = {};
+};
+
+using DynamicScratchpadImpl = DynamicScratchpadImplT<
+      mm::memory_kind::host,
+      mm::memory_kind::pinned,
+      mm::memory_kind::device,
+      mm::memory_kind::managed>;
+
+}  // namespace detail
+
+class DynamicScratchpad
+  : public Scratchpad
+  , private detail::DynamicScratchpadImpl {
+ public:
+  explicit DynamicScratchpad(scratch_sizes_t initial_sizes = {},
+                             AccessOrder device_order = cudaStream_t(0),
+                             AccessOrder pinned_dealloc_order = {},
+                             AccessOrder managed_dealloc_order = {}) {
+    initial_sizes_ = initial_sizes;
+    for (auto &s : initial_sizes_) {
+      if (s == 0)
+        s = 4096;
+    }
+    if (!pinned_dealloc_order.has_value())
+      pinned_dealloc_order = device_order;
+    if (!managed_dealloc_order.has_value())
+      managed_dealloc_order = device_order;
+
+    set_upstream_resrouce<mm::memory_kind::host>(mm::GetDefaultResource<mm::memory_kind::host>());
+
+    set_upstream_resrouce<mm::memory_kind::pinned>(
+        mm::GetDefaultResource<mm::memory_kind::pinned>(),
+        AccessOrder::host(),
+        pinned_dealloc_order);
```

Comment:
What would be the use case for other than device ordered deallocation order? Copying the data from GPU to the CPU?

Reply:
Now that I think of it, if we use scratchpad for copy-out, then we can delete device memory in host order, too - unless there are multiple streams involved. In any case, the default is to use device order here (see 10 lines up).